### PR TITLE
[IMP] website, *: allow creating anchor links 

### PR DESCRIPTION
**Pull request by kea14**
* portal, web_editor

Anchor Link:
- Some visual improvements of the link dialog.
- New 'Page Anchor' field.
    - System by default: top and bottom.
    - Shows existing anchors.

Creation of the anchor:
- New option 'Anchor Name' to be able to create an anchor.

task-1913458

### CODE REVIEW

**Comment by kea14** [comment]
PR created. Working on it.

**Comment by kea14** [comment]
@qsm-odoo not finished but almost, after a last check tomorrow morning on input we discussed about

**Comment by kea14** [comment]
@qsm-odoo Don't know if you are already on it but don't forget to review it when possible :smiley:

File: addons/website/static/src/js/editor/editor.js [hunk 1]
```diff
@@ -5,6 +5,10 @@ var weWidgets = require('web_editor.widget');
 var wUtils = require('website.utils');
 
 weWidgets.LinkDialog.include({
+	custom_events: {
+        website_url_chosen: '_onAutocompleteClose',
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/editor.js]: I am spotting an indentation issue

File: addons/website/static/src/js/editor/editor.js [hunk 2]
```diff
@@ -14,5 +18,17 @@ weWidgets.LinkDialog.include({
...
+    //--------------------------------------------------------------------------
+    // Handlers
+    //--------------------------------------------------------------------------
+
+    /**
+     * @private
+     */
+    _onAutocompleteClose: function () {
+		this._adaptPreview();
+		this._onURLInput();
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/editor.js]: Indentation issue

File: addons/website/static/src/js/editor/editor.js [hunk 3]
```diff
@@ -14,5 +18,17 @@ weWidgets.LinkDialog.include({
...
+
+    //--------------------------------------------------------------------------
+    // Handlers
+    //--------------------------------------------------------------------------
+
+    /**
+     * @private
+     */
+    _onAutocompleteClose: function () {
+		this._adaptPreview();
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/editor.js]: Is this needed ?

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 4]
```diff
@@ -1310,6 +1310,72 @@ registry.many2one = SnippetOption.extend({
...
+    openAnchorDialog: function (previewMode, value, $opt) {
+        var self = this;
+        this.anchorDialog = new Dialog(this, {
+            title: _t("Anchor Name"),
+            $content: $(qweb.render('web_editor.dialog.anchor_name', {
+                currentAnchor: self.$target.attr('id'),
+            })),
+            buttons: [
+                {
+                    text: _t("Validate"),
```
[the remark this answers is not in the document]
Comment (kea14) [reply addons/web_editor/static/src/js/editor/snippets.options.js]: Did this because of this in specs

> click on "Anchor Name", opens a modal with one input : Enter the name of the anchor Link. + validate or discard

In my opinion "Save" is also better

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 5]
```diff
@@ -1520,6 +1521,39 @@ var LinkDialog = Dialog.extend({
...
+    /**
+     * @private
+     */
+    _adaptPageAnchor: function () {
+        var self = this;
+        var urlInputValue = this.$('.o_link_dialog_url_input').val();
+        var $pageAnchor = this.$('.o_link_dialog_page_anchor');
+        var isFromWebsite = urlInputValue.startsWith('/');
+        var isEmpty = urlInputValue.length === 0;
+        var anchorsArray = ['', '#top', '#bottom'];
```
[the remark this answers is not in the document]
Comment (kea14) [reply addons/web_editor/static/src/js/widgets/widgets.js]: The code below does not find top or bottom

File: addons/website/static/src/js/content/snippets.animation.js [hunk 6]
```diff
@@ -1080,6 +1080,31 @@ registry._fixAppleCollapse = Animation.extend({
...
+
+    //--------------------------------------------------------------------------
+    // Handlers
+    //--------------------------------------------------------------------------
+
+    /**
+     * @private
+     */
+    _onAnimateClick: function (ev) {
+        var href = $(ev.currentTarget).attr('href').substring(1);
```
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/website/static/src/js/content/snippets.animation.js]: I understood that... but why?

File: addons/website/static/src/js/content/snippets.animation.js [hunk 7]
```diff
@@ -1080,6 +1080,31 @@ registry._fixAppleCollapse = Animation.extend({
...
+        var anchorExists = $.find(href).length > 0;
```
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/website/static/src/js/content/snippets.animation.js]: $.find(href) = $(href) = crash with all the url I try. $() is supposed to search in the DOM... searching an URL... well it crashes :/ "Error: Syntax error, unrecognized expression"

[the remark this answers is not in the document]
Comment (kea14) [reply addons/website/static/src/js/content/snippets.animation.js]: I don't get it, it is working here. But that being said, openAnchorDialog function is not working as expected anymore and I can see this error "Error: Syntax error, unrecognized expression". I'll take a closer look

File: addons/website/static/src/js/content/snippets.animation.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (kea14) [reply addons/website/static/src/js/content/snippets.animation.js]: Because I forgot about the hash :smile: ...

File: addons/website/static/src/js/content/snippets.animation.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (kea14) [reply addons/website/static/src/js/content/snippets.animation.js]: I still don't get what's crashing here

File: addons/website/static/src/js/editor/editor.js (hunk 1 above)
Comment (kea14) [reply addons/website/static/src/js/editor/editor.js]: Enlighten me, I don't see it

File: addons/website/static/src/js/editor/editor.js (hunk 2 above)
Comment (kea14) [reply addons/website/static/src/js/editor/editor.js]: ?

**Comment by kea14** [comment]
@qsm-odoo Updated, you can check when you have time. 
One thing I didn't change : anchorsArray in _adaptPageAnchor

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 8]
```diff
@@ -1310,6 +1310,79 @@ registry.many2one = SnippetOption.extend({
...
+            $content: $(qweb.render('web_editor.dialog.anchorName', {
+                currentAnchor: self.$target.attr('id'),
+            })),
+            buttons: [
+                {
+                    text: _t("Save"),
+                    classes: 'btn-primary',
+                    click: function () {
+                        var $input = this.$('.o_input_anchor_name');
+                        var anchorName = $input.val().trim().split(' ').join('_');
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js]: `$input.val().trim().replace(/\s/g, '_')` is clearer and 60% more efficient

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 9]
```diff
@@ -1310,6 +1310,79 @@ registry.many2one = SnippetOption.extend({
...
+                        var alreadyExists = $('#' + anchorName).length > 0;
+                        $input.toggleClass('is-invalid', !alreadyExists);
+                        if (!alreadyExists) {
+                            self._setAnchorName(anchorName);
+                        } else {
+                            $input.toggleClass('is-invalid');
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js]: Two `toggleClass` of the same class on the same element with opposite condition...

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 10]
```diff
@@ -1310,6 +1310,79 @@ registry.many2one = SnippetOption.extend({
...
+    /**
+     * @private
+     *
+     * @param {String} value
+     * @param {Event} ev
+     */
+    _setAnchorName: function (value) {
+        this.$target.attr('id', value);
+        this.$target.trigger('content_changed');
+        this.anchorDialog.close();
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js]: Should not be this method's job to close the modal and that modal does not need to be saved in a variable.

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 11]
```diff
@@ -1310,6 +1310,79 @@ registry.many2one = SnippetOption.extend({
...
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * @private
+     *
+     * @param {String} value
+     * @param {Event} ev
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js]: ...

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 12]
```diff
@@ -1520,6 +1521,39 @@ var LinkDialog = Dialog.extend({
...
     },
+    /**
+     * @private
+     */
+    _adaptPageAnchor: function () {
+        var self = this;
+        var urlInputValue = this.$('#o_link_dialog_url_input').val();
+        var $pageAnchor = this.$('.o_link_dialog_page_anchor');
+        var isFromWebsite = _.str.startsWith(urlInputValue, '/');
+        var isEmpty = urlInputValue.length === 0;
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/widgets/widgets.js]: No need of that variable and the condition should be written `!urlInputValue.length` (convention)

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 13]
```diff
@@ -1520,6 +1521,39 @@ var LinkDialog = Dialog.extend({
...
+        var anchorsArray = ['', '#top', '#bottom'];
+        var $selectMenu = this.$('select[name="link_anchor"]');
+
+        if (isEmpty) {
+            $selectMenu.prop('selectedIndex', 0);
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/widgets/widgets.js]: Why is this needed ?

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 14]
```diff
@@ -1520,6 +1521,39 @@ var LinkDialog = Dialog.extend({
...
+                $selectMenu.empty();
+                _.each($(response).find(':not(p).oe_structure, :not(p)[data-oe-type=html]').children(), function (el) {
+                    if (typeof el.id !== 'undefined' && el.id.length > 0) {
+                        anchorsArray.push('#' + el.id);
+                    }
+                });
+                _.map(anchorsArray, function (elem) {
+                    $selectMenu.append($("<option></option>").text(elem));
+                });
+            }).done(function () {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/widgets/widgets.js]: Use `always` instead of `done`

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 15]
```diff
@@ -1520,6 +1521,39 @@ var LinkDialog = Dialog.extend({
...
+        if (isEmpty) {
+            $selectMenu.prop('selectedIndex', 0);
+        }
+        $pageAnchor.toggleClass('d-none', !isFromWebsite);
+        if (isFromWebsite) {
+            this.$('.o_anchors_loading').removeClass('d-none');
+            $.get(window.location.origin + urlInputValue, function (response) {
+                $selectMenu.empty();
+                _.each($(response).find(':not(p).oe_structure, :not(p)[data-oe-type=html]').children(), function (el) {
+                    if (typeof el.id !== 'undefined' && el.id.length > 0) {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/widgets/widgets.js]: `ed.id !== undefined`, but in fact this cannot happen, `el.id` is always a string so the condition is useless
=>
```javascript
if (el.id.length) {
```

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 16]
```diff
@@ -1520,6 +1521,39 @@ var LinkDialog = Dialog.extend({
...
+                        anchorsArray.push('#' + el.id);
+                    }
+                });
+                _.map(anchorsArray, function (elem) {
+                    $selectMenu.append($("<option></option>").text(elem));
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/widgets/widgets.js]:
```javascript
$('<option>', {text: elem})
```

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 17]
```diff
@@ -1565,6 +1599,17 @@ var LinkDialog = Dialog.extend({
     // Handlers
     //--------------------------------------------------------------------------
 
+    /**
+     * @private
+     */
+    _onAnchorChange: function () {
+        var anchorValue = this.$('.link-style').val();
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/widgets/widgets.js]: .link-style?

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 18]
```diff
@@ -1565,6 +1599,17 @@ var LinkDialog = Dialog.extend({
...
+        var urlInput = this.$('#o_link_dialog_url_input').val();
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/widgets/widgets.js]: Wrong name

File: addons/web_editor/static/src/xml/editor.xml [hunk 19]
```diff
@@ -267,6 +267,14 @@
...
                     <div class="col-md-9">
                         <input type="text" name="url" class="form-control" id="o_link_dialog_url_input" required="required"/>
+                        <small class="form-text">Hint: Type '/' to link to an existing page.</small>
+                    </div>
+                </div>
+                <div class="form-group row o_link_dialog_page_anchor d-none">
+                    <label class="col-form-label col-md-3" for="o_link_dialog_anchor_input">Page Anchor</label>
+                    <div class="col-md-9">
+                        <select name="link_anchor" class="form-control link-style"></select>
+                        <span class="font-weight-bold o_anchors_loading">Loading...</span>
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/xml/editor.xml]: form text too

File: addons/website/static/src/js/content/snippets.animation.js [hunk 20]
```diff
@@ -1080,6 +1080,31 @@ registry._fixAppleCollapse = Animation.extend({
...
+
+    /**
+     * @private
+     */
+    _onAnimateClick: function (ev) {
+        var hash = window.location.hash;
+        var $anchor = $(hash);
+        if ($anchor.length) {
+            ev.preventDefault();
+            $("html, body").animate({
```
Comment (qsm-odoo) [inline addons/website/static/src/js/content/snippets.animation.js]: `'` instead of `"`

File: addons/website/static/src/js/content/snippets.animation.js [hunk 21]
```diff
@@ -1080,6 +1080,31 @@ registry._fixAppleCollapse = Animation.extend({
...
+    },
+
+    //--------------------------------------------------------------------------
+    // Handlers
+    //--------------------------------------------------------------------------
+
+    /**
+     * @private
+     */
+    _onAnimateClick: function (ev) {
```
Comment (qsm-odoo) [inline addons/website/static/src/js/content/snippets.animation.js]: That method is not correct. Image you have a link `/aboutus#top` on your homepage... with your animation you will prevent the user to go to the aboutus page and scroll the homepage to the top...
Maybe the animation should not target link clicks, but URL changes instead.

**Comment by kea14** [comment]
@qsm-odoo Do not forget to check when possible please :smiley:

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 22]
```diff
@@ -1520,6 +1521,35 @@ var LinkDialog = Dialog.extend({
...
+            this.$('.o_anchors_loading').removeClass('d-none');
+            $.get(window.location.origin + urlInputValue, function (response) {
+                $selectMenu.empty();
+                _.each($(response).find(':not(p).oe_structure, :not(p)[data-oe-type=html]').children(), function (el) {
+                    if (el.id.length > 0) {
+                        anchorsArray.push('#' + el.id);
+                    }
+                });
+                _.map(anchorsArray, function (elem) {
+                    $selectMenu.append($('<option>', {text: elem}));
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/widgets/widgets.js]: Why not pushing the option in selectMenu directly instead of filling a table first then going through it to create the options ? :)

File: addons/web_editor/static/src/xml/editor.xml [hunk 23]
```diff
@@ -474,4 +482,15 @@
...
         </form>
     </div>
+
+    <!-- Anchor Name option dialog -->
+    <div t-name="web_editor.dialog.anchorName">
+        <div class="form-group row">
+            <label class="col-form-label col-md-3" for="anchorName">Anchor Name</label>
+            <div class="col-md-9">
+                <input type="text" class="form-control o_input_anchor_name" id="anchorName" t-attf-value="#{currentAnchor}" placeholder="Anchor name"/>
+                <div class="invalid-feedback o_anchor_name_invalid">The name chosen already exists</div>
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/xml/editor.xml]: chosen name

File: addons/website/static/src/js/content/snippets.animation.js [hunk 24]
```diff
@@ -1080,6 +1080,31 @@ registry._fixAppleCollapse = Animation.extend({
...
+        var hash = window.location.hash;
+        var $anchor = $(hash);
+        if ($anchor.length && this.$target[0].pathname === window.location.pathname) {
```
Comment (qsm-odoo) [inline addons/website/static/src/js/content/snippets.animation.js]: Comment remains: would not it be better to target URL changes instead of link clicks ? There are ways to listen to hash changes I think ( https://caniuse.com/#search=hashchange, using jQuery). That way no need to check the pathname and add a click handler on every link, juste creating an animation targeting the #wrapwrap element and listening to this event on the window should do trick.

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 25]
```diff
@@ -1520,6 +1521,35 @@ var LinkDialog = Dialog.extend({
...
+        var urlInputValue = this.$('#o_link_dialog_url_input').val();
+        var $pageAnchor = this.$('.o_link_dialog_page_anchor');
+        var isFromWebsite = _.str.startsWith(urlInputValue, '/');
+        var anchorsArray = ['', '#top', '#bottom'];
+        var $selectMenu = this.$('select[name="link_anchor"]');
+
+        $pageAnchor.toggleClass('d-none', !isFromWebsite);
+        if (isFromWebsite) {
+            this.$('.o_anchors_loading').removeClass('d-none');
+            $.get(window.location.origin + urlInputValue, function (response) {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/widgets/widgets.js]: @JKE-be What is your opinion on this? I am still unsure (getting the anchors in javascript by fetching the URL). My main concern is that the feature will never be testable this way... but I don't know if creating a route for this is complicated for nothing.

File: addons/web_editor/static/src/js/widgets/widgets.js (hunk 22 above)
Comment (kea14) [reply addons/web_editor/static/src/js/widgets/widgets.js]: That is what I did initially but whitespace, #top and #bottom (from anchorsArray) doesn't show up that's why I decided to fill selectMenu with _.map just after. Do you have an idea to fix that ?

File: addons/website/static/src/js/content/snippets.animation.js (hunk 24 above)
Comment (kea14) [reply addons/website/static/src/js/content/snippets.animation.js]: I tried to handle it in start function with selector #wrapwrap but it doesn't seem to behave like expected. It detects hash change but .animate() isn't working as expected (buggy). What am I missing ?

```suggestion
start: function () {
    $(window).on('hashchange', function (ev) {
        console.log('hashchange');
        var hash = window.location.hash;
        var $anchor = $(hash);
        if ($anchor.length) {
            ev.preventDefault();
            $('html, body').animate({
                scrollTop: $anchor.offset().top,
            }, 500);
        }
    });
},
```

Comment (qsm-odoo) [reply addons/website/static/src/js/content/snippets.animation.js]: @kea14 Then I suppose the body scrolls then the event is fired, so bad idea.

Anyways, please try testing and make sure everything is handled and thought through before sending to review, I easily found two things which are not ok... 
- If the anchor is put in the page then the user tries to make a link to it (most common user case), you cannot. Because $.get will render the not-yet-saved page and thus not found the new anchor... you must have seen this during development !
- First click on an anchor does not scroll with animation.

**Comment by qsm-odoo** [comment]
@kea14 Ready?

**Comment by kea14** [comment]
@qsm-odoo No, I prefer pinging you when it's the right moment for you to have a look at it :smile: 
This said, I fixed what you told me in the last comments but I'm adjusting something else as SBU tested and gave me his feedback.

**Comment by kea14** [comment]
@qsm-odoo Everything seems to work. SBU should test it again and see if everything looks good for him tomorrow morning. I guess you can already check my code, when you have some time available. Thanks

**Comment by kea14** [comment]
@qsm-odoo Could you check when possible ? Still waiting a reply from SBU concerning the task itself but everything is supposed to work

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 26]
```diff
@@ -1310,6 +1310,77 @@ registry.many2one = SnippetOption.extend({
...
+
+    //--------------------------------------------------------------------------
+    // Public
+    //--------------------------------------------------------------------------
+
+    /**
+     * @override
+     */
+    onClone: function () {
+        this.$target.removeAttr('id');
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js]: And removing the data-anchor ?

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 27]
```diff
@@ -1520,6 +1521,41 @@ var LinkDialog = Dialog.extend({
...
+    /**
+     * @private
+     */
+    _adaptPageAnchor: function () {
+        var self = this;
+        var urlInputValue = this.$('input[name="url"]').val();
+        var $pageAnchor = this.$('.o_link_dialog_page_anchor');
+        var isFromWebsite = urlInputValue[0] === '/' || urlInputValue[0] === '#';
+        var $selectMenu = this.$('select[name="link_anchor"]');
+        var anchorsArray = ['', '#top', '#bottom'];
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/widgets/widgets.js]: Not needed anymore and the code below can be adapted and improved

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 28]
```diff
@@ -1520,6 +1521,41 @@ var LinkDialog = Dialog.extend({
...
+            this.$('.o_anchors_loading').removeClass('d-none');
+            var def;
+            if (urlInputValue !== window.location.pathname && urlInputValue !== '#') {
+                def = $.get(window.location.origin + urlInputValue);
+            } else {
+                def = $.when(document.body.outerHTML);
+            }
+            def.then(function (response) {
+                $selectMenu.empty();
+                _.each($(response).find('[data-anchor=true]'), function (el) {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/widgets/widgets.js]: I would write `[id][data-anchor="true"]`

File: addons/web_editor/static/src/js/widgets/widgets.js (hunk 22 above)
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/widgets/widgets.js]: Can be done now

**Comment by kea14** [comment]
@qsm-odoo updated

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 29]
```diff
@@ -1310,6 +1310,77 @@ registry.many2one = SnippetOption.extend({
...
+    //--------------------------------------------------------------------------
+    // Options
+    //--------------------------------------------------------------------------
+
+    /**
+     * @see this.selectClass for parameters
+     */
+    openAnchorDialog: function (previewMode, value, $opt) {
+        var self = this;
+        this.anchorDialog = new Dialog(this, {
```
[the remark this answers is not in the document]
Comment (kea14) [reply addons/web_editor/static/src/js/editor/snippets.options.js]: `this._setAnchorName(anchorName);` won't work. What am I missing ?

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js]: Why would it not work?

[the remark this answers is not in the document]
Comment (kea14) [reply addons/web_editor/static/src/js/editor/snippets.options.js]: `Uncaught TypeError: this._setAnchorName is not a function`

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js]: There is no `this._setAnchorName` in your code ... cache ? local code ? Don't know.

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js]: `this.anchorDialog` is useless ^^' the line I commented ^^

File: addons/website/static/src/js/utils.js [hunk 30]
```diff
@@ -23,12 +23,26 @@ function autocompleteWithPages(self, $input) {
...
-                    return r.loc;
-                });
-                response(rs);
+                if ($input.val()[0] === '#') {
+                    var anchors = [];
+                    var def = $.when(document.body.outerHTML);
+                    def.then(function (response) {
+                        _.each($(response).find('[id][data-anchor="true"]'), function (el) {
+                            anchors.push('#' + el.id);
+                        });
```
[the remark this answers is not in the document]
Comment (kea14) [reply addons/website/static/src/js/utils.js]: I block a bit on this one, if you have one more hint for me when you have time @qsm-odoo

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/website/static/src/js/utils.js]: This code is a duplicate of what you implemented directly for the anchor input. What you should do is share that code. What I have in mind is an `options` parameter to this `autocompleteWithPages` function so that it proposes pages or anchor or both. Then the URL input in the link dialog will use both functionalities and the anchor input will use anchor only (by calling this `autocompleteWithPages` function with the right parameters when initializing the `<input/>`s) . That way your code is reusable and more importantly not duplicated.

**Comment by kea14** [comment]
@qsm-odoo check it again when you have some time :smiley: thanks

**Comment by kea14** [comment]
@qsm-odoo Changes are done

File: addons/website/static/src/js/utils.js [hunk 31]
```diff
@@ -15,19 +34,28 @@ var qweb = core.qweb;
...
-                args: [null, request.term],
-                kwargs: {
-                    limit: 15,
-                },
-            }).then(function (exists) {
-                var rs = _.map(exists, function (r) {
-                    return r.loc;
+            if (request.term[0] === '#') {
+                loadAnchors(request.term).then(function (anchors) {
+                    response(anchors);
```
Comment (qsm-odoo) [inline addons/website/static/src/js/utils.js]: Should not you filter the anchors according to the user term here ? Or is the spec explicitly says it should be like this ?

Comment (kea14) [reply addons/website/static/src/js/utils.js]: Well, specs do not seem to be strict on this part. I guess the user should see at least every existing anchors when he types '#'.

File: addons/web_editor/static/src/js/widgets/widgets.js [hunk 32]
```diff
@@ -1574,10 +1576,12 @@ var LinkDialog = Dialog.extend({
...
      */
-    _onURLInput: function (ev) {
-        $(ev.currentTarget).closest('.form-group').removeClass('o_has_error').find('.form-control, .custom-select').removeClass('is-invalid');
-        var isLink = $(ev.currentTarget).val().indexOf('@') < 0;
+    _onURLInput: function () {
+        var $linkUrlInput = this.$('#o_link_dialog_url_input');
+        $linkUrlInput.closest('.form-group').removeClass('o_has_error').find('.form-control, .custom-select').removeClass('is-invalid');
+        var isLink = $linkUrlInput.val().indexOf('@') < 0;
         this.$('input[name="is_new_window"]').closest('.form-group').toggleClass('d-none', !isLink);
+        this._adaptPageAnchor();
```
[the remark this answers is not in the document]
Comment (kea14) [reply addons/web_editor/static/src/js/widgets/widgets.js]: So I should also create _onURLInput function for this in editor.js, right ?

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/widgets/widgets.js]: Yes, extending it. Just calling _super then this I suppose.

File: addons/website/static/src/js/utils.js (hunk 31 above)
Comment (qsm-odoo) [reply addons/website/static/src/js/utils.js]: Fine for now

File: addons/web_editor/static/src/js/widgets/widgets.js (hunk 32 above)
[the remark this answers is not in the document]
Comment (kea14) [reply addons/web_editor/static/src/js/widgets/widgets.js]: Is this ok as updated ?

**Comment by kea14** [comment]
@qsm-odoo updated

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 33]
```diff
@@ -1310,6 +1310,82 @@ registry.many2one = SnippetOption.extend({
     }
 });
 
+/**
+ * Handles the edition of snippet's anchor name.
+ */
+registry.anchorName = SnippetOption.extend({
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js]: All of this should be in website too.

File: addons/web_editor/static/src/xml/editor.xml [hunk 34]
```diff
@@ -267,6 +267,14 @@
...
                     <div class="col-md-9">
                         <input type="text" name="url" class="form-control" id="o_link_dialog_url_input" required="required"/>
+                        <small class="form-text">Hint: Type '/' to link to an existing page and '#' to an anchor.</small>
+                    </div>
+                </div>
+                <div class="form-group row o_link_dialog_page_anchor d-none">
+                    <label class="col-form-label col-md-3" for="o_link_dialog_anchor_input">Page Anchor</label>
+                    <div class="col-md-9">
+                        <select name="link_anchor" class="form-control link-style"></select>
+                        <small class="form-text font-weight-bold o_anchors_loading">Loading...</small>
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/xml/editor.xml]: All of this should be in website too.

File: addons/web_editor/static/src/xml/editor.xml [hunk 35]
```diff
@@ -474,4 +482,15 @@
...
+    <!-- Anchor Name option dialog -->
+    <div t-name="web_editor.dialog.anchorName">
+        <div class="form-group row">
+            <label class="col-form-label col-md-3" for="anchorName">Anchor Name</label>
+            <div class="col-md-9">
+                <input type="text" class="form-control o_input_anchor_name" id="anchorName" t-attf-value="#{currentAnchor}" placeholder="Anchor name"/>
+                <div class="invalid-feedback">The name chosen already exists</div>
+            </div>
+        </div>
+    </div>
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/xml/editor.xml]: All of this should be in website too.

File: addons/website/static/src/js/editor/editor.js [hunk 36]
```diff
@@ -5,6 +5,14 @@ var weWidgets = require('web_editor.widget');
 var wUtils = require('website.utils');
 
 weWidgets.LinkDialog.include({
+    events: _.extend({}, weWidgets.LinkDialog.prototype.events, {
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/editor.js]: `|| {}` for the case where events is undefined one day

File: addons/website/static/src/js/editor/editor.js [hunk 37]
```diff
@@ -5,6 +5,14 @@ var weWidgets = require('web_editor.widget');
...
+        'change select[name="link_anchor"]': '_onAnchorChange',
+        'input input[name="url"]': '_onURLInput',
+    }),
+    custom_events: _.extend({}, weWidgets.LinkDialog.prototype.events, {
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/editor.js]: I let you find what is wrong here.

File: addons/website/static/src/js/editor/editor.js [hunk 38]
```diff
@@ -14,5 +22,62 @@ weWidgets.LinkDialog.include({
...
+        var anchorValue = this.$('[name="link_anchor"]').val();
+        var $urlInput = this.$('[name="url"]');
+        var urlInputValue = $urlInput.val();
+        if (urlInputValue.indexOf('#') > -1) {
+            urlInputValue = urlInputValue.substr(0, urlInputValue.indexOf('#'));
+        }
+        $urlInput.val(urlInputValue + anchorValue);
+    },
+    /**
+     * @private
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/editor.js]: `@override`

File: addons/website/static/src/js/utils.js [hunk 39]
```diff
@@ -6,6 +6,25 @@ var core = require('web.core');
 
 var qweb = core.qweb;
 
+/**
+ * Allows to load anchors from a page.
+ *
+ * @param {string} url
```
Comment (qsm-odoo) [inline addons/website/static/src/js/utils.js]: \+ `@returns {Deferred<string[]>}`

**Comment by kea14** [comment]
@qsm-odoo Updated :smiley:

File: addons/web_editor/static/src/xml/editor.xml [hunk 40]
```diff
@@ -267,8 +267,10 @@
                     <label class="col-form-label col-md-3" for="o_link_dialog_url_input">URL or Email</label>
                     <div class="col-md-9">
                         <input type="text" name="url" class="form-control" id="o_link_dialog_url_input" required="required"/>
+                        <small class="form-text">Hint: Type '/' to link to an existing page and '#' to an anchor.</small>
                     </div>
                 </div>
+                <t t-call="website.page_anchor_field"/>
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/xml/editor.xml]: Those two are not right... You should know it :/ You write "website." in a web_editor template :/ If you do not know something, look in the doc (which is updated for that kind of stuff) and if you do not find anything, ask me directly.

Here there you not be anything. All you have to do is an extension template in the website app, thanks to `t-extend`

File: addons/website/static/src/js/editor/editor.js [hunk 41]
```diff
@@ -5,6 +5,14 @@ var weWidgets = require('web_editor.widget');
 var wUtils = require('website.utils');
 
 weWidgets.LinkDialog.include({
+    events: _.extend({}, weWidgets.LinkDialog.prototype.events || {}, {
+        'change select[name="link_anchor"]': '_onAnchorChange',
+        'input input[name="url"]': '_onURLInput',
+    }),
+    custom_events: _.extend({}, weWidgets.LinkDialog.prototype.events || {}, {
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/editor.js]: You did not adapt here :/ You write `events` inside `custom_events`...

Comment (kea14) [reply addons/website/static/src/js/editor/editor.js]: Sorry focused on adding `|| {}`

File: addons/web_editor/static/src/xml/editor.xml [hunk 42]
```diff
@@ -263,10 +263,11 @@
                         <input type="text" name="label" class="form-control" id="o_link_dialog_label_input" required="required" t-att-value="widget.data.text"/>
                     </div>
                 </div>
-                <div class="form-group row">
+                <div class="form-group row o_url_input">
                     <label class="col-form-label col-md-3" for="o_link_dialog_url_input">URL or Email</label>
                     <div class="col-md-9">
                         <input type="text" name="url" class="form-control" id="o_link_dialog_url_input" required="required"/>
+                        <small class="form-text">Hint: Type '/' to link to an existing page and '#' to an anchor.</small>
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/xml/editor.xml]: That hint is still not true without website...

File: addons/website/static/src/js/editor/editor.js [hunk 43]
```diff
@@ -5,6 +5,14 @@ var weWidgets = require('web_editor.widget');
 var wUtils = require('website.utils');
 
 weWidgets.LinkDialog.include({
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/editor.js]: In fact add the website xml file as dependencies otherwise the t-extend may not be applied in all cases.

**Comment by qsm-odoo** [comment]
@robodoo r+
Thanks @kea14
